api/src/test/java/org/pmiops/workbench/api/AdminAuthorityTest.java

```diff
@@ -0,0 +1,197 @@
+   * convention. Ideally these REST paths should eventually be updated for consistency. Avoid adding
+   * more entries to this list and put /admin/ in the REST path instead.
+   */
+  private static final Set<String> NONCONFORMING_ADMIN_METHODS =
+      ImmutableSet.of(
+          "postStatusAlert",
+          "publishWorkspace",
+          "unpublishWorkspace",
+          "getInstitution",
+          "getInstitutions",
+          "createInstitution",
+          "deleteInstitution",
+          "updateInstitution",
+          "setInstitutionUserInstructions",
+          "deleteInstitutionUserInstructions",
+          "createAuthDomain",
+          "updateUserDisabledStatus");
+
+  /**
+   * Conversely, methods which have an "/admin/" infix, but do not require an Authority. This should
+   * be very exceptional, and should entail some additional method of authentication.
+   *
+   * <p>Please add a comment for any exceptions added to this list.
+   */
+  private static final Set<String> NONAUTHORITY_ADMIN_METHODS =
+      ImmutableSet.of(
+          // logEgressEvent requires an API key secret; it's a webhook which cannot use authority
+          "logEgressEvent",
+          // Not technically an admin endpoint, but only enabled in lower environments
+          "unsafeSelfBypassAccessRequirement");
+
+  /**
+   * Heuristic test: if there's a REST endpoint that looks like an admin endpoint (based on the
+   * URL), ensure that it has an @AuthorityRequired annotation. Assert the inverse as well.
+   * Exceptions are enumerated in the above allowlists.
+   *
+   * <p>To achieve this, we use reflection over our API controller methods as well as the Swagger
+   * generated interfaces. There are three relevant classes:
+   *
+   * <ul>
+   *   <li>our source code controller implementation, which may or may not have @AuthorityRequired
+   *   <li>the generated FooApiDelegate interface, which our controller implements
+   *   <li>and the generated FooApi interface, which integrates with Spring, and contains an
+   *       annotation @RequestMapping with request path information
+   * </ul>
+   *
+   * <p>We walk these relationships to check our heuristic. The following preconditions must hold:
+   *
+   * <ul>
+   *   <li>all API controllers live in the RW "api" Java package
+   *   <li>Swagger codegen uses the naming pattern of FooApiDelegate and FooApi
+   * </ul>
+   */
+  @Test
+  public void testAdminEndpointsHaveAuthority() throws Exception {
+    List<Class<?>> controllers = findClassesInApiPackageWithAnnotation(RestController.class);
+    Map<String, Class<?>> generatedApis =
+        findClassesInApiPackageWithAnnotation(Api.class).stream()
+            .collect(Collectors.toMap(Class::getName, Functions.identity()));
+
+    assertThat(controllers).isNotEmpty();
+    assertThat(generatedApis).isNotEmpty();
+
+    Map<Class<?>, Class<?>> controllerToApi = new HashMap<>();
+    for (Class controller : controllers) {
```

Comment:
can be `Class<?>` but I'm not sure if that is a real improvement.  Applies elsewhere too.

Reply:
Yes, I'm not totally sure on style here - I thought just a naked Class used to be a Java warning.. these two representations seem equivalent to me `Class`  vs `Class<?>`, though they are definitely distinct types as seen by the compiler. Anyways, for now I'll make them all consistent with `Class<?>`, didn't look much deeper.